### PR TITLE
adds a basic pytest structure for running ocs tests

### PR DESCRIPTION
**Pull request by andrewschoen**
This adds a ``conftest.py`` file that will inspect the test path at collection time and give appropriate markers if the test file lives in a directory with the name of a ceph daemon type. For example, all tests that live in a file at ``tests/mon/test_mons.py`` would get a ``mons`` marker. This is really helpful when wanting to run daemon specific tests.

This also includes a simple tox.ini file that can be used to run pytest against the ``ocs-ci/tests`` folder.

### CODE REVIEW

File: tests/tox.ini [hunk 1]
```diff
@@ -0,0 +1,11 @@
+[tox]
+envlist = py27,py35,py36,py37
```
Comment (vasukulkarni) [inline tests/tox.ini]: @andrewschoen we have moved out of py27 and anyway its just couple of months for EOL for 27, Can we remove py27?

Comment (andrewschoen) [reply tests/tox.ini]: Sure, I can do that. I ran this with ``py37`` locally anyway.

File: tests/tox.ini [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+envlist = py35,py36,py37
+skip_missing_interpreters = true
+skipsdist = true
+
+[testenv]
+deps=
+  -rrequirements.txt
+  pytest
+commands=
+    pytest -v {posargs:.}
```
Comment (clacroix12) [inline tests/tox.ini]: I checked [the docs](https://tox.readthedocs.io/en/latest/example/general.html) but wasn't able to find anything on it, would you elaborate what the `.` is doing in `{posargs:.}`?

File: tests/tox.ini [hunk 3]
```diff
@@ -0,0 +1,11 @@
+[tox]
+envlist = py35,py36,py37
```
Comment (clacroix12) [inline tests/tox.ini]: I don't think we can use py35 here since we utilize f-strings which were introduced in 3.6.

File: tests/tox.ini (hunk 2 above)
Comment (andrewschoen) [reply tests/tox.ini]: It's just running pytest in the current directory, totally not necessary I suppose. I find ``{posargs}`` useful though, if by some reason we ever need to point at a different test directory interactively.

Would you prefer if I remove that default value there? If it improves readability I'm ok with that.

File: tests/tox.ini (hunk 3 above)
Comment (andrewschoen) [reply tests/tox.ini]: I can remove it. Honestly the python version here isn't all that terribly important. If we only want ``ocs-ci`` to support ``py36`` or ``py37`` we can pick one and then name the tox environment something more descriptive than the python version. That's probably be a good idea anyway.

File: tests/tox.ini (hunk 2 above)
Comment (clacroix12) [reply tests/tox.ini]: Ah I see now, that makes sense. For some reason I was thinking it was something for tox but you're just passing `.` through to pytest. That makes sense as the default functionality to me.

File: tests/tox.ini (hunk 3 above)
Comment (clacroix12) [reply tests/tox.ini]: Yeah that sounds like a better approach. If we have multiple envs in `envlist` won't we end up executing tests once per env? I don't think that's what we want to happen for our ocs tests.

Comment (andrewschoen) [reply tests/tox.ini]: I've renamed the env ``ocs-tests`` and defaulted to ``py37``. I'm open to suggestions on better naming for sure, but I think that works. You can specify which env you want to run with the ``-e`` flag so I don't think that'd be a problem.

File: tests/tox.ini [hunk 4]
```diff
@@ -0,0 +1,11 @@
+[tox]
+envlist = {ocs-tests}
+basepython = py37
+skipsdist = true
+
+[testenv]
+deps=
+  -rrequirements.txt
```
Comment (vavuthu) [inline tests/tox.ini]: is that typo ( -r requirements.txt )

File: tests/tox.ini [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+  pytest
```
Comment (petr-balogh) [inline tests/tox.ini]: I see pytest is defined just here in tox requirements. But if we are going support pytest in OCS-CI, we will be implementing some plugins and so on I think that it should be requirements defined for whole ocs-ci project in setup.py.

File: tests/tox.ini (hunk 4 above)
Comment (petr-balogh) [reply tests/tox.ini]: it's not typo @vavuthu , if you will do pip install -rrequirements.txt it work like a char as just -r is the first argument and then the path to requirement file following. It's more readable when there is space I agree, but doesn't matter.

Comment (petr-balogh) [reply tests/tox.ini]: Do we really need to separate and have other tox.ini file in tests directory? Cannot we just have different env defined in tox.ini file in root directory of the project?

File: tests/requirements.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+-e ../
```
Comment (petr-balogh) [inline tests/requirements.txt]: What about ppl who would like to run pytest. Currently now the flow is like create venv for ocs-ci like 
```bash
python3 -m venv .venv
source .venv/bin/activate
pip install -r requirements.txt
python run.py ......
```

If I would like to run pytest now, I don't have defined the pytest requirement anywhere just for tox.
So if we will stick with this project, I would have one place for all the requirements in setup.py as everything will be binded so closely with whole framework I think.  Or are you tending to have separate requirement file for tests requirement, so once ppl will need to have some package to be installed for their test that it should be defined in this file and not in the setup.py?  If so I still think that at least pytest and it's plugins should be in the setup.py.

Also as the user if I would like to install this requirements to the same venv as I have for ocs-ci I cannot do something like this from root directory of the project 
```bash
pip install -r tests/requirements.txt
Directory '../' is not installable. File 'setup.py' not found.
```
I have to `cd tests` and run `pip install -r requirements.txt`.

So please document a bit how this should be used for the users.

File: tests/tox.ini [hunk 7]
```diff
@@ -0,0 +1,11 @@
+[tox]
```
Comment (mbukatov) [inline tests/tox.ini]: What is the purpose of this file here?

File: tests/tox.ini (hunk 2 above)
Comment (mbukatov) [reply tests/tox.ini]: This is also useful if one needs to pass any other pytest option to pytest run executed via tox, eg. like `tox -e py36 -- --pdb --runxfail tests/test_foo.py`.

That said, I don't see any value in using tox here to run ocsqe test cases.

File: tests/requirements.txt (hunk 6 above)
Comment (mbukatov) [reply tests/requirements.txt]: +1 I don't see any value in using tox here to run ocsqe test cases.

File: tests/tox.ini (hunk 5 above)
Comment (andrewschoen) [reply tests/tox.ini]: Good point, I'll make that change.

**Comment by andrewschoen** [comment]
I do like to tox to run these tests, but I'm not sold on it if the rest of the team disagrees. Even if we keep the ``tests/tox.ini`` file those who chose to can run pytest directly. I've used it on past projects to great success, especially the generative envlist feature. It can be used to create a dynamic testing matrix very easily. https://tox.readthedocs.io/en/latest/config.html#generating-environments-conditional-settings

And yes, this definitely needs some documentation. Fully agree there.

File: tests/requirements.txt (hunk 6 above)
Comment (andrewschoen) [reply tests/requirements.txt]: I've added ``pytest`` to setup.py and added a comment below about tox usage.

File: tests/tox.ini (hunk 4 above)
Comment (andrewschoen) [reply tests/tox.ini]: We could have a different env defined in the tox.ini file in the root. With as simple as it is now, that might be a good option. I've found in the past though that multiple, less complicated tox.ini files are better than trying to manage one very complicated tox.ini that does everything.

**Comment by RazTamir** [comment]
@clacroix12 @petr-balogh @vasukulkarni is this PR still relevant?

**Comment by clacroix12** [comment]
@RazTamir i don't think so. We are handling marks differently with petr's PR and it seems like we have decided to execute pytest directly instead of through tox.

File: tests/conftest.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    daemon in it, tag the test with the appropriate
+    marker. This allows for us to filter and test
+    by specific daemon type.
+
+    For example, all tests that live in a file at
+    tests/mon/test_mons.py would be tagged with
+    a 'mons' marker.
+    """
+    for item in items:
+        test_path = item.location[0]
```
Comment (petr-balogh) [inline tests/conftest.py]: Hi @andrewschoen, what item.location[0] contains? First directory in which the test file belongs to?  We discussed to have structure starting with teams folders in QE. And we merged PR #61 which implements pytest for us and the basic config we had in old runner to pytest. So if this structure is needed from engineering can this been discussed on our regular meeting? Thanks

**Comment by clacroix12** [comment]
@petr-balogh @RazTamir can we close this PR given the recent changes we have made with regards to pytest?

**Comment by andrewschoen** [comment]
> @petr-balogh @RazTamir can we close this PR given the recent changes we have made with regards to pytest?

I'll go ahead and close this. I can reopen it if we decide there's things here we might still need.

**Pull request closed without merging**